### PR TITLE
Hex package diff lock package

### PR DESCRIPTION
**Pull request by xinz**

### CODE REVIEW

**Comment by xinz** [comment]
Related https://github.com/hexpm/hex/pull/705

File: lib/mix/tasks/hex.package.ex [hunk 1]
```diff
@@ -20,9 +20,14 @@ defmodule Mix.Tasks.Hex.Package do
...
-  ## Fetch and diff package contents between versions
+  ## Diff package versions
+
+      mix hex.package diff APP VERSION
+
+  This command compares the project's dependency `APP` against
+  the target package version, unpacking the target version into
+  temporary directory and running a diff command.
 
-      mix hex.package diff PACKAGE VERSION1 VERSION2
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.package.ex:25]: @xinz this was probably one of the conflicts, can we revert this change? we should keep support for `mix hex.package diff PACKAGE VERSION1 VERSION2`.

Btw, thank you so much for looking into it!

Comment (xinz) [reply lib/mix/tasks/hex.package.ex:25]: hello @wojtekmach, thanks, that's added now, please let me know if any missing.

Comment (wojtekmach) [reply lib/mix/tasks/hex.package.ex:25]: Sorry for confusion, this wasn’t the only line that needed keeping, we need to ensure we don’t accidentally break that now existing feature.

Comment (xinz) [reply lib/mix/tasks/hex.package.ex:25]: Got it, I will double check the changed files between the current `main` branch and the origin PR's changed files, thanks for your clarification.

Comment (xinz) [reply lib/mix/tasks/hex.package.ex:25]: Hi @wojtekmach, base on the `main` branch to check the original changed files, there were some missed have been reverted, please review again when you are free, thanks!

File: lib/mix/tasks/hex.package.ex [hunk 2]
```diff
@@ -92,18 +100,15 @@ defmodule Mix.Tasks.Hex.Package do
...
-        diff(repo(opts), package, version_range)
+      ["diff", package, version] ->
+        diff(repo(opts), package, parse_version!(version))
 
       _ ->
         Mix.raise("""
           Invalid arguments, expected one of:
 
           mix hex.package fetch PACKAGE VERSION [--unpack]
-          mix hex.package diff PACKAGE VERSION1 VERSION2
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.package.ex:106]: please revert this change

File: lib/mix/tasks/hex.package.ex [hunk 3]
```diff
@@ -113,8 +118,8 @@ defmodule Mix.Tasks.Hex.Package do
   def tasks() do
     [
       {"fetch PACKAGE VERSION [--unpack]", "Fetch the package"},
-      {"diff PACKAGE VERSION1 VERSION2", "Fetch and diff package contents between versions"},
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.package.ex:116]: please revert this change

File: test/mix/tasks/hex.package_test.exs [hunk 4]
```diff
@@ -95,23 +151,27 @@ defmodule Mix.Tasks.Hex.PackageTest do
...
+  after
+    purge([ReleaseDeps.MixProject])
   end
 
   test "diff: success (variant args)" do
-    in_tmp(fn ->
+    in_diff_fixture(fn ->
       Hex.State.put(:diff_command, "git diff --no-index --no-color __PATH1__ __PATH2__")
 
-      assert catch_throw(Mix.Tasks.Hex.Package.run(["diff", "ex_doc", "0.0.1", "0.1.0"])) ==
```
Comment (wojtekmach) [inline test/mix/tasks/hex.package_test.exs:104]: please revert this change

File: lib/mix/tasks/hex.package.ex [hunk 5]
```diff
@@ -203,61 +213,96 @@ defmodule Mix.Tasks.Hex.Package do
...
         version2 = Hex.Version.parse!(version2)
         {to_string(version1), to_string(version2)}
 
-      _ ->
-        Mix.raise(
-          "Expected version range to be in format `VERSION1..VERSION2`, got: `#{inspect(string)}`"
-        )
+      [version] ->
+        version = Hex.Version.parse!(version)
+        to_string(version)
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.package.ex:305]:
```suggestion
        version |> Hex.Version.parse!() |> to_string()
```

File: lib/mix/tasks/hex.package.ex (hunk 2 above)
Comment (xinz) [reply lib/mix/tasks/hex.package.ex:106]: thanks, done.

File: lib/mix/tasks/hex.package.ex (hunk 3 above)
Comment (xinz) [reply lib/mix/tasks/hex.package.ex:116]: thanks, done, I revert this line back, and keep the description the same refer the changes from https://github.com/hexpm/hex/pull/705/commits/47d7949a3137bae96d282707be35fa9cf8b3484a .

File: test/mix/tasks/hex.package_test.exs (hunk 4 above)
Comment (xinz) [reply test/mix/tasks/hex.package_test.exs:104]: thanks, done.

File: lib/mix/tasks/hex.package.ex (hunk 5 above)
Comment (xinz) [reply lib/mix/tasks/hex.package.ex]: thanks, done.

**Comment by xinz** [comment]
@wojtekmach thanks, I misunderstood something in the previous, now it is much clear.

File: test/mix/tasks/hex.package_test.exs [hunk 6]
```diff
@@ -120,27 +180,27 @@ defmodule Mix.Tasks.Hex.PackageTest do
       assert_received {:mix_shell, :run, [out]}
       assert out =~ "hex_metadata.config\nmix.exs"
     end)
-  end
-
-  test "diff: bad version range" do
-    msg = "Expected version range to be in format `VERSION1..VERSION2`, got: `\"1.0.0..\"`"
-
-    assert_raise Mix.Error, msg, fn ->
-      Mix.Tasks.Hex.Package.run(["diff", "ex_doc", "1.0.0.."])
```
Comment (wojtekmach) [inline test/mix/tasks/hex.package_test.exs:129]: I think we should keep this test (and associated behaviour) but if it makes implementation simpler I think it's fine to leave it as is. Given we have two ways of specifying version range, `x..y` and `x y`, perhaps we should deprecate the former anyway (given the latter is more unix friendly) but that's a separate discussion.

**Review by wojtekmach — APPROVED** [review]
Thank you!

**Comment by wojtekmach** [comment]
Thank you @RyanSiu1995 and @xinz, really appreciated!
